File: 01.py
#!/usr/bin/env python3

from aochelper import *

filename = sys.argv[1] if len(sys.argv) > 1 else sys.argv[0].replace('py', 'txt')
inp = open(filename).read().splitlines()
words = ['one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine']

def solve(part2=False):
    ans = 0
    for line in inp:
        digit = ''
        for i in range(len(line)):
            if not ('0' <= line[i] <= '9'):
                if not part2: continue
                for j, num in enumerate(words):
                    if line[i:].startswith(num):
                        digit += str(j+1)
            else:
                digit += line[i]
        if digit:
            ans += int(digit[0] + digit[-1])
    return ans

print(solve(), solve(True))

File: 01.rs
use std::fs::File;
use std::io::{self, BufRead};

fn p1(inp: &str) -> i32 {
    let d: Vec<_> = inp.chars().filter(|c| c.is_digit(10)).collect();
    format!("{}{}", d[0], d[d.len() - 1]).parse().unwrap()
}

fn p2(inp: &str) -> i32 {
    let mut inp = inp.to_owned();
    let words = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"];

    for (i, n) in words.iter().enumerate().map(|(i, n)| (i + 1, *n)) {
        inp = inp.replace(n, &format!("{}{}{}", n.chars().next().unwrap(), i, n.chars().last().unwrap()));
    }

    p1(&inp)
}

fn main() {
    if let Ok(file) = File::open("01.txt") {
        let lines: Vec<_> = io::BufReader::new(file).lines().filter_map(Result::ok).collect();

        let mut sum1 = 0;
        let mut sum2 = 0;

        for line in &lines {
            sum1 += p1(line);
            sum2 += p2(line);
        }

        println!("{} {}", sum1, sum2);
    }
}
